src/winrtble/peripheral.rs

```diff
@@ -81,13 +86,66 @@ impl Peripheral {
             properties.manufacturer_data = manufacturer_data
                 .into_iter()
                 .map(|d| {
-                    let company_id = d.company_id().unwrap();
+                    let manufacturer_id = d.company_id().unwrap();
                     let buffer = d.data().unwrap();
                     let reader = DataReader::from_buffer(&buffer).unwrap();
                     let len = reader.unconsumed_buffer_length().unwrap() as usize;
                     let mut data = vec![0u8; len];
                     reader.read_bytes(&mut data).unwrap();
-                    (company_id, data)
+
+                    // Emit event of newly received advertisement
+                    self.adapter.emit(CentralEvent::ManufacturerAdvertisement {
+                        address: self.address,
+                        manufacturer_id,
+                        data: data.clone(),
+                    });
+
+                    (manufacturer_id, data)
+                })
+                .collect();
+        }
+
+        if let Ok(data_sections) = advertisement.data_sections() {
+            properties.service_data = data_sections
+                .into_iter()
+                .filter_map(|d| {
+                    let buffer = d.data().unwrap();
+                    let reader = DataReader::from_buffer(&buffer).unwrap();
+                    let len = reader.unconsumed_buffer_length().unwrap() as usize;
+                    let mut data = vec![0u8; len];
+                    reader.read_bytes(&mut data).unwrap();
+
+                    let service_data = match d.data_type().unwrap() {
+                        AdvertisementDataType::SERVICE_DATA_16_BIT_UUID => {
+                            let (uuid, data) = data.split_at(2);
+                            let uuid = uuid_from_u16(u16::from_le_bytes(uuid.try_into().unwrap()));
+                            Some((uuid, data.to_owned()))
+                        }
+                        AdvertisementDataType::SERVICE_DATA_32_BIT_UUID => {
+                            let (uuid, data) = data.split_at(4);
+                            let uuid = uuid_from_u32(u32::from_le_bytes(uuid.try_into().unwrap()));
+                            Some((uuid, data.to_owned()))
+                        }
+                        AdvertisementDataType::SERVICE_DATA_128_BIT_UUID => {
+                            let (uuid, data) = data.split_at(16);
+                            let uuid = Uuid::from_slice(uuid).unwrap();
+                            Some((uuid, data.to_owned()))
+                        }
+                        _ => None,
```

Comment:
If you `return None` here, then you can avoid the if let immediately after.

Reply:
That was a good idea, Thank you!